### DISCUSSION TITLE
PDF to Image without scale factor losing Quality.

### DISCUSSION THREAD

**Discussion by Ric-Wolf** [pypdfium2-team/pypdfium2#319] [Q&A] [answered]
Im using the library for 2 reasons. I need to convert a pdf to a jpg and also crop the pdf before the conversion.
All examples use with the render method the kwarg scale=4. but this is giving me a file 4 times the size (in pixels) I need the image file to have the same size and without losing quality. Is there an option to not compress the bitmap or a pillow option to increase the quality. On to_pil im using quality=100

**Comment by mara004** [comment]
Am I right to assume the pages of your PDF are in fact embedded JPEG images?
In that case, you could use [`PdfImage.extract()`](https://pypdfium2.readthedocs.io/en/stable/python_api.html#pypdfium2._helpers.pageobjects.PdfImage.extract), which should be lossless for JPEG/JPX.
Further, `jpegtran` may support lossless JPEG cropping, according to its description.

However, note that `PdfImage.extract()` will lose the original encoding for non-JPEG filters.
In general, I recommend to use pikepdf for image extraction, it's _much_ better at this task than pdfium.

**Reply by Ric-Wolf** [reply]
I'm actually not sure if the PDF is in fact an image, I use an API to transform an html string I wrote into the PDF but that API returns the PDF in "LEGAL" but I need it in "A4" thats the crop part. Then I need to save the exact PDF as an image file.
I will check if the extract returns everything as an image because the API I use to convert to PDF is not documented but I cannot use any other technology (company rules)

**Reply by mara004** [reply, marked as answer]
If the PDF pages in question do not consist of just an embedded image each, then there may be some fundamental misunderstanding in the first post.

What then do you mean by "I need the image file to have the same size and without losing quality"? PDFs are vector and images are raster, so rendering is _by design_ a lossy operation that will create a bitmap of a certain size in pixels, and thereby a fixed displaying quality. Rendered bitmaps are bound to be bigger than the vector source in terms of file size.

> On to_pil im using quality=100

`to_pil()` does not take any such argument:
https://github.com/pypdfium2-team/pypdfium2/blob/ff2ff4bd519e9b44bfe3182cff052d893bd442e0/src/pypdfium2/_helpers/bitmap.py#L210

If you mean PIL's `save()` method, and are using JPEG, that is, of course, a lossy encoding (quality=100 is just "good quality" and not "lossless"). If you want to save the pixel information exactly as provided by pdfium, then don't use JPEG. Use e.g. PNG or PPM instead.

**Reply by mara004** [reply]
BTW, if you need to keep a reference to paper size in the rendered image, you can set DPI metadata (`bidi_dpi = 72*scale`).
I think you can pass `dpi = (bidi_dpi, bidi_dpi)` to PIL's `.save()`.
Conversely, to get a rendering scale from DPI input, you can do `dpi / 72`, because 1 pdf canvas unit is 1/72in (aside from `UserUnit`, which pdfium does not support yet).
